evidence/verify.go

```
@@ -1,79 +1,257 @@
-				evidence.Time(),
-				header.Time,
-			)
-		}
+	// check that the evidence isn't already committed
+	if evpool.isCommitted(evidence) {
+		return nil, errors.New("evidence was already committed")
+	}
+
+	// verify the time of the evidence
+	blockMeta := evpool.blockStore.LoadBlockMeta(evidence.Height())
+	if blockMeta == nil {
+		return nil, fmt.Errorf("don't have header at height #%d", evidence.Height())
 	}
+	evTime := blockMeta.Header.Time
+	ageDuration := state.LastBlockTime.Sub(evTime)
 
+	// check that the evidence hasn't expired
 	if ageDuration > evidenceParams.MaxAgeDuration && ageNumBlocks > evidenceParams.MaxAgeNumBlocks {
-		return fmt.Errorf(
+		return nil, fmt.Errorf(
 			"evidence from height %d (created at: %v) is too old; min height is %d and evidence can not be older than %v",
 			evidence.Height(),
-			evidence.Time(),
+			evTime,
 			height-evidenceParams.MaxAgeNumBlocks,
 			state.LastBlockTime.Add(evidenceParams.MaxAgeDuration),
 		)
 	}
 
-	valset, err := stateDB.LoadValidators(evidence.Height())
+	// apply the evidence-specific verification logic
+	switch ev := evidence.(type) {
+	case *types.DuplicateVoteEvidence:
+		valSet, err := evpool.stateDB.LoadValidators(evidence.Height())
+		if err != nil {
+			return nil, err
+		}
+		err = VerifyDuplicateVote(ev, state.ChainID, valSet)
+		if err != nil {
+			return nil, fmt.Errorf("verifying duplicate vote evidence: %w", err)
+		}
+
+		_, val := valSet.GetByAddress(ev.VoteA.ValidatorAddress)
+
+		return &Info{
+			Evidence:         evidence,
+			Time:             evTime,
+			Validators:       []*types.Validator{val}, // just a single validator for duplicate vote evidence
+			TotalVotingPower: valSet.TotalVotingPower(),
+		}, nil
+
+	case *types.LightClientAttackEvidence:
+		commonHeader, err := getSignedHeader(evpool.blockStore, evidence.Height())
+		if err != nil {
+			return nil, err
+		}
+		commonVals, err := evpool.stateDB.LoadValidators(evidence.Height())
```

same question

---

The load validators function should return an error when commonVals is nil